dns-320l_GPL/shadow-4.0.18.2/man/sulogin.8
.\"     Title: sulogin
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.73.2 <http://docbook.sf.net/>
.\"      Date: 10/28/2007
.\"    Manual: System Management Commands
.\"    Source: System Management Commands
.\"
.TH "SULOGIN" "8" "10/28/2007" "System Management Commands" "System Management Commands"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
sulogin - Single-user login
.SH "SYNTAX"
.PP
\fBsulogin\fR
[\fItty\-device\fR]
.SH "DESCRIPTION"
.PP
The
\fBsulogin\fR
command is invoked by
\fBinit\fR
prior to allowing the user access to the system when in single user mode\. This feature may only be available on certain systems where
\fBinit\fR
has been modified accordingly, or where the
\fI/etc/inittab\fR
has an entry for a single user login\.
.PP
The user is prompted
.PP
Type control\-d to proceed with normal startup,
(or give root password for system maintenance):
.PP
Input and output will be performed with the standard file descriptors unless the optional device name argument is provided\.
.PP
If the user enters the correct root password, a login session is initiated\. When
\fIEOF\fR
is pressed instead, the system enters multi\-user mode\.
.PP
After the user exits the single\-user shell, or presses
\fIEOF\fR, the system begins the initialization process required to enter multi\-user mode\.
.SH "CAVEATS"
.PP
This command can only be used if
\fBinit\fR
has been modified to call
\fBsulogin\fR
instead of
\fI/bin/sh\fR, or if the user has set the
\fIinittab\fR
to support a single user login\. For example, the line:
.PP
co:s:respawn:/etc/sulogin /dev/console
.PP
should execute the sulogin command in single user mode\.
.PP
As complete an environment as possible is created\. However, various devices may be unmounted or uninitialized and many of the user commands may be unavailable or nonfunctional as a result\.
.SH "FILES"
.PP
\fI/etc/passwd\fR
.RS 4
User account information\.
.RE
.PP
\fI/etc/shadow\fR
.RS 4
Secure user account information\.
.RE
.SH "SEE ALSO"
.PP

\fBlogin\fR(1),
\fBsh\fR(1),
\fBinit\fR(8)\.


dns-320l_GPL/tinylogin-1.4/sulogin.c
/* vi: set sw=4 ts=4: */
#include <fcntl.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <termios.h>
#include <unistd.h>
#include <utmp.h>
#include <sys/resource.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/types.h>
#include <ctype.h>
#include <time.h>

#include "tinylogin.h"


// sulogin defines
#define SULOGIN_PROMPT "\nGive root password for system maintenance\n" \
	"(or type Control-D for normal startup):"

static const char *forbid[] = {
	"ENV",
	"BASH_ENV",
	"HOME",
	"IFS",
	"PATH",
	"SHELL",
	"LD_LIBRARY_PATH",
	"LD_PRELOAD",
	"LD_TRACE_LOADED_OBJECTS",
	"LD_BIND_NOW",
	"LD_AOUT_LIBRARY_PATH",
	"LD_AOUT_PRELOAD",
	"LD_NOWARN",
	"LD_KEEPDIR",
	(char *) 0
};



static void catchalarm()
{
	exit(EXIT_FAILURE);
}


extern int sulogin_main(int argc, char **argv)
{
	char *cp;
	char *device = (char *) 0;
	const char *name = "root";
	int timeout = 0;
	static char pass[BUFSIZ];
	struct termios termio;
	struct passwd pwent;
	struct passwd *pwd;
	time_t start, now;
	const char **p;
#ifdef CONFIG_FEATURE_SHADOWPASSWDS
	struct spwd *spwd = NULL;
#endif							/* CONFIG_FEATURE_SHADOWPASSWDS */

	tcgetattr(0, &termio);
	/* set control chars */
	termio.c_cc[VINTR]  = 3;	/* C-c */
	termio.c_cc[VQUIT]  = 28;	/* C-\ */
	termio.c_cc[VERASE] = 127; /* C-? */
	termio.c_cc[VKILL]  = 21;	/* C-u */
	termio.c_cc[VEOF]   = 4;	/* C-d */
	termio.c_cc[VSTART] = 17;	/* C-q */
	termio.c_cc[VSTOP]  = 19;	/* C-s */
	termio.c_cc[VSUSP]  = 26;	/* C-z */
	/* use line dicipline 0 */
	termio.c_line = 0;
	/* Make it be sane */
	termio.c_cflag &= CBAUD|CBAUDEX|CSIZE|CSTOPB|PARENB|PARODD;
	termio.c_cflag |= CREAD|HUPCL|CLOCAL;
	/* input modes */
	termio.c_iflag = ICRNL | IXON | IXOFF;
	/* output modes */
	termio.c_oflag = OPOST | ONLCR;
	/* local modes */
	termio.c_lflag = ISIG | ICANON | ECHO | ECHOE | ECHOK | ECHOCTL | ECHOKE | IEXTEN;
	tcsetattr(0, TCSANOW, &termio);
	openlog("sulogin", LOG_PID | LOG_CONS | LOG_NOWAIT, LOG_AUTH);
	if (argc > 1) {
		if (strncmp(argv[1], "-t", 2) == 0) {
			if (strcmp(argv[1], "-t") == 0) {
				if (argc > 2) {
					timeout = atoi(argv[2]);
					if (argc > 3) {
						device = argv[3];
					}
				}
			} else {
				if (argc > 2) {
					device = argv[2];
				}
			}
		} else {
			device = argv[1];
		}
		if (device) {
			close(0);
			close(1);
			close(2);
			if (open(device, O_RDWR) >= 0) {
				dup(0);
				dup(0);
			} else {
				syslog(LOG_WARNING, "cannot open %s\n", device);
				exit(EXIT_FAILURE);
			}
		}
	}
	if (access(passwd_file, 0) == -1) {
		syslog(LOG_WARNING, "No password file\n");
		error_msg_and_die("No password file\n");
	}
	if (!isatty(0) || !isatty(1) || !isatty(2)) {
		exit(EXIT_FAILURE);
	}


	/* Clear out anything dangerous from the environment */
	for (p = forbid; *p; p++)
		unsetenv(*p);


	signal(SIGALRM, catchalarm);
	alarm(timeout);
	if (!(pwd = getpwnam(name))) {
		syslog(LOG_WARNING, "No password entry for `root'\n");
		error_msg_and_die("No password entry for `root'\n");
	}
	pwent = *pwd;
#ifdef CONFIG_FEATURE_SHADOWPASSWDS
	spwd = NULL;
	if (pwd && ((strcmp(pwd->pw_passwd, "x") == 0)
				|| (strcmp(pwd->pw_passwd, "*") == 0))) {
		endspent();
		spwd = getspnam(name);
		if (spwd) {
			pwent.pw_passwd = spwd->sp_pwdp;
		}
	}
#endif							/* CONFIG_FEATURE_SHADOWPASSWDS */
	while (1) {
		cp = getpass(SULOGIN_PROMPT);
		if (!cp || !*cp) {
			puts("\n");
			fflush(stdout);
			syslog(LOG_INFO, "Normal startup\n");
			exit(EXIT_SUCCESS);
		} else {
			safe_strncpy(pass, cp, sizeof(pass));
			bzero(cp, strlen(cp));
		}
		if (strcmp(pw_encrypt(pass, pwent.pw_passwd), pwent.pw_passwd) == 0) {
			break;
		}
		time(&start);
		//now = start;
		time( &now );
		//while (difftime(now, start) < FAIL_DELAY) {
		while ( (now-start) < FAIL_DELAY) {
			sleep(FAIL_DELAY);
			time(&now);
		}
		puts("Login incorrect");
		fflush(stdout);
		syslog(LOG_WARNING, "Incorrect root password\n");
	}
	bzero(pass, strlen(pass));
	alarm(0);
	signal(SIGALRM, SIG_DFL);
	puts("Entering System Maintenance Mode\n");
	fflush(stdout);
	syslog(LOG_INFO, "System Maintenance Mode\n");
	run_shell(pwent.pw_shell, 1, 0, 0);
	return (0);
}
